controllers/mirrorpeersecret_controller.go

```diff
@@ -0,0 +1,282 @@
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	MetaSecretName    = "mirrorpeer-meta-secret"
+	PeerNamesKey      = "peer-names"
+	InternalNamespace = "odfmo-system"
+)
+
+// MirrorPeerSecretReconciler reconciles a MirrorPeer CRs,
+// and source/destination Secrets
+type MirrorPeerSecretReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=multicluster.odf.openshift.io,resources=mirrorpeers,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=multicluster.odf.openshift.io,resources=mirrorpeers/status,verbs=get
+//+kubebuilder:rbac:groups=core,resources=pods;secrets;configmaps;events,verbs=*
+//+kubebuilder:rbac:groups=rbac.authorization.k8s.io,resources=roles;rolebindings,verbs=*
+//+kubebuilder:rbac:groups=certificates.k8s.io,resources=certificatesigningrequests;certificatesigningrequests/approval,verbs=get;list;watch;create;update
+//+kubebuilder:rbac:groups=certificates.k8s.io,resources=signers,verbs=approve
+//+kubebuilder:rbac:groups=cluster.open-cluster-management.io,resources=managedclusters,verbs=get;list;watch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+
+// Reconcile standard reconcile function for MirrorPeerSecret controller
+func (r *MirrorPeerSecretReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	// if there is no 'namespace', it must be a 'MirrorPeer' object
+	if req.Namespace == "" {
+		return r.mirrorPeerReconcile(ctx, req)
+	}
+	// all other cases, it should be 'Secret' (soucre|BLUE or destination|GREEN)
+	return r.mirrorPeerSecretReconcile(ctx, req)
+}
+
+func (r *MirrorPeerSecretReconciler) mirrorPeerReconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	var mirrorPeerObj multiclusterv1alpha1.MirrorPeer
+	err := r.Get(ctx, req.NamespacedName, &mirrorPeerObj)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			/*
+				what all actions to be taken when a MirrorPeer CR deleted?
+			*/
+			err = nil // neglect any not found error
+		}
+		if err != nil {
+			logger.Error(err, "[MirrorPeerSecret] Error while retrieving 'MirrorPeer' object", "controller", "MirrorPeerSecret")
+		}
+		return ctrl.Result{}, err
+	}
+	logger.Info("[MirrorPeerSecret] Found a MirrorPeer Object", "controller", "MirrorPeerSecret", "MirrorPeer", mirrorPeerObj)
```

Comment:
Why do we need to check for MirrorPeer object availablilty here. Shouldn't the original reconcile func handle this ?

Reply:
Right... Moved the changes to the MirrorPeer reconciler.